treegp/kernels.py

```diff
@@ -0,0 +1,392 @@
+            if invLam is not None:
+                raise TypeError("Cannot set both invLam and scale_length in AnisotropicVonKarman.")
+            invLam = np.diag(1./np.array(scale_length)**2)
+
+        self.ndim = invLam.shape[0]
+        self.ntheta = self.ndim*(self.ndim+1)//2
+        self._d = np.diag_indices(self.ndim)
+        self._t = np.tril_indices(self.ndim, -1)
+        self.set_params(invLam)
+        bounds = np.array(bounds)
+        if bounds.ndim == 1:
+            bounds = np.repeat(bounds[None, :], self.ntheta, axis=0)
+        assert bounds.shape == (self.ntheta, 2)
+        self._bounds = bounds
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        from scipy.spatial.distance import pdist, cdist, squareform
+        from scipy import special
+        X = np.atleast_2d(X)
+
+        if Y is None:
+            dists = pdist(X, metric='mahalanobis', VI=self.invLam)
+            Filter = (dists != 0.)
+            K = np.zeros_like(dists)
+            K[Filter] = dists[Filter] **(5./6.) *  special.kv(5./6., 2*np.pi * dists[Filter])
+            lim0 = special.gamma(5./6.) /(2 * ((np.pi)**(5./6.)) )
+            K = squareform(K)
+            np.fill_diagonal(K, lim0)
+            K /= lim0
+        else:
+            if eval_gradient:
+                raise ValueError(
+                    "Gradient can not be evaluated.")
+            dists = cdist(X, Y, metric='mahalanobis', VI=self.invLam)
+            Filter = (dists != 0.)
+            K = np.zeros_like(dists)
+            K[Filter] = dists[Filter] **(5./6.) *  special.kv(5./6., 2*np.pi * dists[Filter])
+            lim0 = special.gamma(5./6.) /(2 * ((np.pi)**(5./6.)) )
+            if np.sum(Filter) != len(K[0])*len(K[:,0]):
+                K[~Filter] = lim0
+            K /= lim0
+
+        if eval_gradient:
+            raise ValueError(
+                "Gradient can not be evaluated.")
+        else:
+            return K
+
+    @property
+    def hyperparameter_cholesky_factor(self):
+        return Hyperparameter("CholeskyFactor", "numeric", (1e-5, 1e5), int(self.ntheta))
+
+    def get_params(self, deep=True):
+        return {"invLam":self.invLam}
+
+    def set_params(self, invLam=None):
+        if invLam is not None:
+            self.invLam = invLam
+            self._L = np.linalg.cholesky(self.invLam)
+            self._theta = np.hstack([np.log(self._L[self._d]), self._L[self._t]])
+
+    @property
+    def theta(self):
+        return self._theta
+
+    @theta.setter
+    def theta(self, theta):
+        self._theta = theta
+        self._L = np.zeros_like(self.invLam)
+        self._L[np.diag_indices(self.ndim)] = np.exp(theta[:self.ndim])
+        self._L[np.tril_indices(self.ndim, -1)] = theta[self.ndim:]
+        self.invLam = np.dot(self._L, self._L.T)
+
+    def __repr__(self):
+        return "{0}(invLam={1!r})".format(self.__class__.__name__, self.invLam)
+
+    @property
+    def bounds(self):
+        return self._bounds
```

Comment:
This code is redundant with code in the anisotropic RBF kernel. I think we want to implement an `AnisotropicKernel`base class from which both the `AnisotropicRBF` and `AnisotropicVonKarman` kernels both inherit these methods.

Reply:
This is  totally correct. I guess the problem is the `sklearn` API that does not allowed for anisotropic kernel and this is the main issue. In `george`, they implemented the RBF kernel, and after you have to specify the number of dimension that allowed for anisotropy. You can imagine in the future to make for `treegp` its own API for kernel.